src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,70 +283,241 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
         return decrypted;
     }
 
-    private char getEncodeVersionLengthPrefix() {
-        return (char)('a' + ENCODE_VERSION.length());
-    }
-
     /**
-     * encrypt text with current key based on API level
+     * Get Secret Key based on API level to use in encryption. Decryption key
+     * depends on version# since user can migrate to new Android.OS
      *
-     * @param clearText
      * @return
+     * @throws NoSuchAlgorithmException
+     */
+    public synchronized SecretKey loadSecretKeyForEncryption() throws IOException, GeneralSecurityException {
+        // Loading key only once for performance. If API is upgraded, it will
+        // restart the device anyway. It will load the correct key for new API.
+        if (mKey != null && mMacKey != null)
+            return mKey;
+
+        final byte[] secretKeyData = AuthenticationSettings.INSTANCE.getSecretKeyData();
+        if(secretKeyData == null && Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            throw new IllegalStateException("Secret key must be provided for API < 18. " +
+                    "Use AuthenticationSettings.INSTANCE.setSecretKey()");
+        }
+        
+        if (secretKeyData != null) {
+            mBlobVersion = VERSION_USER_DEFINED;
+        } else {
+            mBlobVersion = VERSION_ANDROID_KEY_STORE;
+        }
+        
+        return getKeyOrCreate(mBlobVersion);
+    }
+
+    /**
+     * load key based on version for migration
+     * 
      * @throws GeneralSecurityException
-     * @throws UnsupportedEncodingException
+     * @throws IOException
      */
-    public String encrypt(String clearText)
-            throws GeneralSecurityException, IOException {
+    public synchronized SecretKey loadSecretKeyForDecryption(String keyVersion) throws GeneralSecurityException, IOException {
+        if (mSecretKeyFromAndroidKeyStore != null) {
+            return mSecretKeyFromAndroidKeyStore;
+        }
+        
+        return getKey(keyVersion);
+    }
 
-        Logger.v(TAG, "Starting encryption");
+    /**
+     * For API <18 or user provide the key, will return the user supplied key.
+     * Supported API >= 18 PrivateKey is stored in AndroidKeyStore. Loads key
+     * from the file if it exists. If not exist, it will generate one.
+     * @param keyVersion
+     * @return
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    private synchronized SecretKey getKeyOrCreate(final String keyVersion) throws GeneralSecurityException, IOException {
+        if (VERSION_USER_DEFINED.equals(keyVersion)) {
+            return getKey(keyVersion);
+        }
+        
+        try {
+            mSecretKeyFromAndroidKeyStore = getKey(keyVersion);
+        } catch (final IOException | GeneralSecurityException exception) {
+            // If encountering exception for reading keys, try to generate new keys
```

Comment:
Move this out of the catch block. 
Check if mSecretKeyFromAndroidKeyStore is null, and generate the key.